### PR TITLE
[BugFix] remove nil inserts in models

### PR DESCRIPTION
**Pull request by 6543**
## this blocks #9084

### CODE REVIEW

File: modules/migrations/gitea.go [hunk 1]
```diff
@@ -393,18 +393,18 @@ func (g *GiteaLocalUploader) CreateIssues(issues ...*base.Issue) error {
...
+			return err
+		}
 		g.issues.Store(is.Index, is.ID)
 	}
 	return nil
 }
 
 // CreateComments creates comments of issues
 func (g *GiteaLocalUploader) CreateComments(comments ...*base.Comment) error {
+
```
Comment (6543) [inline modules/migrations/gitea.go]:
```suggestion
```

File: models/migrate.go [hunk 2]
```diff
@@ -151,8 +156,10 @@ func InsertIssueComments(comments []*Comment) error {
 			reaction.IssueID = comment.IssueID
 			reaction.CommentID = comment.ID
 		}
-		if _, err := sess.Insert(comment.Reactions); err != nil {
-			return err
+		if len(comment.Reactions) != 0 {
```
Comment (zeripath) [inline models/migrate.go]:
```suggestion
		if len(comment.Reactions) > 0 {
```

File: models/migrate.go [hunk 3]
```diff
@@ -196,12 +203,11 @@ func InsertReleases(rels ...*Release) error {
...
 		}
 
-		for i := 0; i < len(rel.Attachments); i++ {
+		for i, a := range rel.Attachments {
 			rel.Attachments[i].ReleaseID = rel.ID
-		}
-
-		if _, err := sess.NoAutoTime().Insert(rel.Attachments); err != nil {
-			return err
+			if _, err := sess.NoAutoTime().Insert(a); err != nil {
```
Comment (zeripath) [inline models/migrate.go]: This is not completely the same. I expect that Insert([]Attachment) will actually attempt a single insert with all the values in it whereas this code will create a separate insert statement for each attachement.

I think you need to switch this back.

File: modules/migrations/gitea.go [hunk 4]
```diff
@@ -393,11 +393,10 @@ func (g *GiteaLocalUploader) CreateIssues(issues ...*base.Issue) error {
 		iss = append(iss, &is)
 	}
 
-	err := models.InsertIssues(iss...)
-	if err != nil {
-		return err
-	}
 	for _, is := range iss {
+		if err := models.InsertIssues(is); err != nil {
```
Comment (zeripath) [inline modules/migrations/gitea.go]: As above

**Comment by 6543** [comment]
@zeripath done

**Comment by guillep2k** [comment]
🤔 I must be missing something, here. Since xorm should ignore empty arrays on insert, what bug is this PR fixing? (I understand it might be faster doing the `if` on this side and skip the xorm call, however).

**Comment by 6543** [comment]
@guillep2k could be a xorm bug but if I remove this cheks migration freeze or xorm throw error: "No element on slice when insert" this is due to xorm update to 1.1...

**Comment by 6543** [comment]
@lunny should explain this better - but it breaks migration at least combined with gitlab migration pull

**Comment by lunny** [comment]
Before xorm v1.0, empty slice when inserting will be ignored and returned successful, but now it will return an error. So users have to check the length of the slice itself.

**Review by lunny — APPROVED** [review]

**Comment by 6543** [comment]
@guillep2k / @zeripath would be nice if you can approve this (or tell me what I have to change)
since I need this pull :)

**Review by lafriks — APPROVED** [review]

**Pull request merged**